src/xalloc.c

```diff
@@ -60,67 +71,276 @@ static long find_insertion_index(void *allocated_ptrs[], long allocated_count, v
+{
+    if (new_size == 0 || old_ptr == NULL)
+    {
+        fprintf(stderr, "arena_xrealloc: invalid arguments (old_ptr=%p, new_size=%zu)\n", old_ptr, new_size);
+        abort();
+    }
+
+    long idx = find_ptr(arena, old_ptr);
+    void *p = realloc(old_ptr, new_size);
+    if (!p)
+    {
+        if (!arena->rollback_in_progress)
+            longjmp(arena->rollback_point, 1);
+        return NULL;
+    }
+    if (idx >= 0)
+        remove_ptr_at(arena, idx);
+    insert_ptr(arena, p);
+
+    return p;
+}
+
+char *arena_xstrdup(arena_t *arena, const char *s)
+{
+    if (s == NULL)
+    {
+        fprintf(stderr, "arena_xstrdup: NULL pointer passed\n");
+        abort();
+    }
+
+    char *p = strdup(s);
+    if (!p)
+    {
+        if (!arena->rollback_in_progress)
+            longjmp(arena->rollback_point, 1);
+        return NULL;
+    }
+    insert_ptr(arena, p);
+    return p;
+}
+
+void arena_xfree(arena_t *arena, void *p)
+{
+    if (!p)
+        return;
+
+    long idx = find_ptr(arena, p);
+    if (idx < 0)
+    {
+        fprintf(stderr, "arena_xfree: double free or corruption detected (%p)\n", p);
+        abort();
+    }
+
+    remove_ptr_at(arena, idx);
+    free(p);
+}
+
+void arena_init_ex(arena_t *arena)
+{
+    // Free existing allocations if arena was previously initialized and still has allocations
+    // We check both allocated_ptrs and allocated_count to avoid false positives from uninitialized memory
+    if (arena->allocated_ptrs != NULL && arena->allocated_count > 0)
+    {
+        arena_reset_ex(arena, false);
+    }
+    
+    arena->rollback_in_progress = false;
+    arena->initial_cap = ARENA_INITIAL_CAP;
+    arena->max_allocations = ARENA_MAX_ALLOCATIONS;
+    arena->allocated_ptrs = calloc(arena->initial_cap, sizeof(void *));
+    if (!arena->allocated_ptrs)
+    {
+        fprintf(stderr, "arena_init_ex: failed to allocate initial pointer array\n");
+        abort();
+    }
+    arena->allocated_cap = arena->initial_cap;
+    arena->allocated_count = 0;
+}
+
+void arena_reset_ex(arena_t *arena, bool verbose)
+{
+    arena->rollback_in_progress = true;
+
+    // Free from the end to avoid expensive memmove on every removal
+    long count = arena->allocated_count;
+    while (arena->allocated_count > 0)
+    {
+        void *p = arena->allocated_ptrs[--arena->allocated_count];
+        free(p);
+    }
+
+    free(arena->allocated_ptrs);
+    arena->allocated_ptrs = NULL;
+    if (verbose)
+    {
+        fprintf(stderr, "Arena reset: freeing %ld allocated blocks\n", count);
+    }
+    arena->allocated_cap = arena->allocated_count = 0;
+
+    arena->rollback_in_progress = false;
+}
+
+void arena_end_ex(arena_t *arena)
+{
+#ifdef DEBUG
+    arena_reset_ex(arena, true);
+#else
+    arena_reset_ex(arena, false);
+#endif
 }
 
 // -------------------------------------------------------------
-// Public API
+// Legacy public API - uses global rollback variables

```

Comment:
This isn't a "legacy" API. It is the normal API.  The 'arena_xmalloc' etc API is just provided for unit testing